### PR TITLE
Add personas and contexts to DB

### PR DESCRIPTION
**Pull request by ChristianSi**
Implements #318. Backend-code only, work in progress.

### CODE REVIEW

File: thentos-core/src/Thentos/Backend/Core.hs [hunk 1]
```diff
@@ -133,6 +133,12 @@ thentosErrorInfo other e = f e
         (Nothing, err400, "thentos session not found")
     f NoSuchServiceSession =
         (Nothing, err400, "service session not found")
+    f NoSuchPersona =
+        (Nothing, err400, "persona not found")
+    f NoSuchContext =
+        (Nothing, err400, "context not found")
+    f MultiplePersonasPerContext =
+        (Nothing, err400, "Cannot register multiple personas by a user for the same context")
```
Comment (fisx) [inline thentos-core/src/Thentos/Backend/Core.hs]: All other error messages start with a small letter.  And I would mention sybil attacks here:

"sybil attack prevention: cannot register multiple personas of the same user for the same context"

Comment (ChristianSi) [inline thentos-core/src/Thentos/Backend/Core.hs]: Will be changed in the next commit. Nevertheless, "Sybil" is a proper name, therefore the error message will still start with an uppercase letter ;)

**Comment by ChristianSi** [comment]
Should be OK for reviewing + merging, if somebody wants to do that. Tests are still incomplete but may be added in another PR.

File: thentos-core/src/Thentos/Types.hs [hunk 2]
```diff
@@ -211,6 +215,38 @@ instance Aeson.FromJSON GroupNode where parseJSON = Aeson.gparseJson
...
+    deriving (Eq, Ord, Show, Read, FromJSON, ToJSON, Typeable, Generic, IsString, FromField,
+              ToField)
+
+newtype ContextDescription = ContextDescription { fromContextDescription :: ST }
+    deriving (Eq, Ord, Show, Read, FromJSON, ToJSON, Typeable, Generic, IsString, FromField,
+              ToField)
+
+data Context = Context
+  { _contextId           :: ContextId
+  , _contextOwnerService :: ServiceId
```
Comment (fisx) [inline thentos-core/src/Thentos/Types.hs]: Rename to `_contextService`?

File: thentos-core/src/Thentos/Util.hs [hunk 3]
```diff
@@ -86,16 +83,6 @@ readsPrecEnumBoundedShow _ s = f [minBound..]
         s' = show x
 
 
--- | Path concatenation for avoiding double slashes in paths.  One
```
Comment (fisx) [inline thentos-core/src/Thentos/Util.hs]: Why was this moved?  And why the trivial extra sub-functions?  I would prefer to undo this; if you disagree, the sub-functions should at least be marked with the INLINE pragma (even though that may be happening automatically).

File: thentos-core/src/Thentos/Transaction.hs [hunk 4]
```diff
@@ -268,6 +269,93 @@ unregisterUserFromService uid sid = void $
...
+-- Connect a persona with a context. Throws an error if the persona is already registered for the
+-- context or if the user has any *other* persona registered for the context. (As we currently
+-- allow only one persona per user and context.)
+registerPersonaWithContext :: Persona -> ContextId -> ThentosQuery e ()
+registerPersonaWithContext persona cxtId = do
+    -- Check that user has no registered personas yet
+    res <- queryT [sql| SELECT count(*)
+                        FROM personas pers, personas_per_context pc
+                        WHERE pers.id = pc.persona_id AND pers.uid = ? AND pc.context_id = ? |]
+                  (Only $ persona ^. personaUid)
```
Comment (fisx) [inline thentos-core/src/Thentos/Transaction.hs]: I think the `context_id = ?` is too much?  Or how are there two `?` and only one argument?

**Comment by fisx** [comment]
Is `Persona`:`process` 1:1, i.e. every context has *exactly* one persona?  Is the global context (empty path in `ProxyUri`) implicitly always there?  Or explicitly?

**Comment by fisx** [comment]
What about `MultipleContextsPerPersona` in `registerPersonaWithContext`?

**Comment by fisx** [comment]
I think I found a runtime error, which convinced me that you should probably complete the tests. (-:

File: thentos-core/src/Thentos/Util.hs (hunk 3 above)
Comment (ChristianSi) [inline thentos-core/src/Thentos/Util.hs]: I created the `stripLeadingSlash` and `stripTrailingSlash` helper functions because I need one of them in `renderProxyUri`. I cannot import Util into Types without an import loop and it felt strange to have the helpers in one module and the operator using them in another, hence I moved everything into Types.

As for the INLINE pragma, I would tend to let the compiler worry about that, except in cases where we found a performance bottleneck via benchmarking. Also, is there even a performance difference between a top-level definition and one defined in a where clause?

**Comment by ChristianSi** [comment]
@fisx:

> Is `Persona`:`process` 1:1, i.e. every context has *exactly* one persona? 

It's 1:n, a user may use the same persona for multiple contexts. But...

> What about `MultipleContextsPerPersona` in `registerPersonaWithContext`?

... currently I don't allow a user to register multiple persons for the same context (Sybil attack), hence that error if they try.

> Is the global context (empty path in `ProxyUri`) implicitly always there?  Or explicitly?

That's up to the application to decide.

> I think I found a runtime error, which convinced me that you should probably complete the tests. (-:

Hey, what did you expect? SQL is dynamically typed, of course it doesn't work unless tested ;) But that's fine, I hadn't "officially" asked for a review and planned to complete the tests anyway. Will do that now.

File: thentos-core/src/Thentos/Transaction.hs (hunk 4 above)
Comment (ChristianSi) [inline thentos-core/src/Thentos/Transaction.hs]: Indeed, the context_id seems spurious.

File: thentos-core/src/Thentos/Types.hs (hunk 2 above)
Comment (ChristianSi) [inline thentos-core/src/Thentos/Types.hs]: I think `_contextOwnerService` is clearer but admittedly it's also a big longish. OK, I'll change it.

**Comment by ChristianSi** [comment]
I completed the transaction tests now; should be good for a final review and possibly merge.

**Comment by fisx** [comment]
looks good.

@ChristianSi could you check if you want to act on any of my last comments?  When done, you can merge.

**Comment by ChristianSi** [comment]
@fisx:

> perhaps it would even be useful to have a function that takes a table name and a number of expected rows in that table and returns a `Spec`?

Good idea, I'll add that to the test utilities.

**Comment by ChristianSi** [comment]
There now is a `rowCountShouldBe` test function and I also refactoring the existing test code to use it.

**Comment by ChristianSi** [comment]
Travis failed with one of its random Selenium timeouts, merging anyway.
